Ensure that OriginalXSymbols are copied over when enabling nullability.

333fred opened the pull request:
Fixes https://github.com/dotnet/roslyn/issues/37879, https://devdiv.visualstudio.com/DevDiv/_workitems/edit/125970/. Previously, original symbols would not be preserved when rewriting for nullability, because these were ad-hoc fields that weren't part of BoundNodes.xml file. As a result, `CSharpSemanticModel.GetMethodGroupSemanticSymbols` thought that overload resolution succeeded when it actually failed, resulting in returning bad symbols from the compiler layer. This change puts these symbol lists into the actual xml, so they're copied during Update. These fields are only necessary for the semantic model, so they don't need to be updated in later compiler phases. To avoid changing massive amounts of code, I looked through the existing usages of the constructors and Update methods, fixed the ones from initial binding, and then added compat shims. @dotnet/roslyn-compiler for review. /cc @jasonmalinowski.

jcouv commented:
Do you want to target `release/dev16.4-preview1` branch?

333fred commented:
Nope, I think we should be taking this for 16.3. It can cause InvalidOperationExceptions and trigger the "A VS process has crashed, please restart VS gold bar."

agocke commented:
fyi @jaredpar I think we should take this for P9

jaredpar commented:
@agocke agree

333fred commented:
@dotnet/roslyn-compiler please review.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:
```diff
@@ -137,17 +137,6 @@ public override Symbol ExpressionSymbol
             }
         }
-
-        /// <summary>
```
><summary> [](start = 12, length = 9)

Consider preserving this information as a comment in the BoundNodes.xml

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:
```diff
@@ -229,17 +218,6 @@ public override Symbol ExpressionSymbol
             get { return this.Indexer; }
         }
 
-        /// <summary>
```
><summary> [](start = 12, length = 9)

Consider preserving this information as a comment in the BoundNodes.xml

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:
```diff
@@ -255,11 +233,11 @@ public BoundIndexerAccess Update(bool useSetterForDefaultArgumentGeneration)
                    this.ArgsToParamsOpt,
                    this.BinderOpt,
                    useSetterForDefaultArgumentGeneration,
+                   this.OriginalIndexersOpt,
```
>this.OriginalIndexersOpt, [](start = 19, length = 25)

Could we replace current implementation of this method with an invocation of a standard Update method?

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:
```diff
@@ -320,17 +298,6 @@ public override Symbol ExpressionSymbol
             get { return this.MethodOpt; }
         }
-
-        /// <summary>
```
><summary> [](start = 12, length = 9)

Consider preserving this information as a comment in the BoundNodes.xml

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:
```diff
@@ -339,17 +306,6 @@ public override Symbol ExpressionSymbol
             get { return this.LogicalOperator; }
         }
-
-        /// <summary>
```
><summary> [](start = 12, length = 9)

Consider preserving this information as a comment in the BoundNodes.xml

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:
```diff
@@ -363,17 +319,6 @@ public override Symbol ExpressionSymbol
             get { return this.MethodOpt; }
         }
-
-        /// <summary>
```
><summary> [](start = 12, length = 9)

Consider preserving this information as a comment in the BoundNodes.xml

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:
```diff
@@ -382,17 +327,6 @@ public override Symbol ExpressionSymbol
             get { return this.MethodOpt; }
         }
-
-        /// <summary>
```
><summary> [](start = 12, length = 9)

Consider preserving this information as a comment in the BoundNodes.xml

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:
```diff
@@ -401,17 +335,6 @@ public override Symbol ExpressionSymbol
             get { return this.Operator.Method; }
         }
-
-        /// <summary>
```
><summary> [](start = 12, length = 9)

Consider preserving this information as a comment in the BoundNodes.xml

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:
```diff
@@ -449,25 +372,14 @@ public override Symbol ExpressionSymbol
             get { return this.SymbolOpt; }
         }
 
-        /// <summary>
```
><summary> [](start = 12, length = 9)

Consider preserving this information as a comment in the BoundNodes.xml

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs:
```diff
@@ -2217,5 +2217,35 @@ public void M(string s)
+            var select = root.DescendantNodes().OfType<IdentifierNameSyntax>().Where(i => i.Identifier.ValueText == "Select").Single();
+            var symbolInfo = model.GetSymbolInfo(select);
+
+            Assert.Null(symbolInfo.Symbol);
```
>Assert.Null(symbolInfo.Symbol); [](start = 12, length = 31)

What did we get before the change?

333fred replied:
We returned a bad symbol that had a `null` `ContainingType`, which the IDE was throwing on.

---
In reply to: [314005550](https://github.com/dotnet/roslyn/pull/37964#discussion_r314005550) [](ancestors = 314005550)

AlekseyTs replied:
> We returned a bad symbol that had a null ContainingType, which the IDE was throwing on.

What was that symbol and where it was coming from? Was it an error MethodSymbol from the bound node?

---
In reply to: [314007924](https://github.com/dotnet/roslyn/pull/37964#discussion_r314007924) [](ancestors = 314007924,314005550)

333fred replied:
Yes, exactly.

---
In reply to: [314009018](https://github.com/dotnet/roslyn/pull/37964#discussion_r314009018) [](ancestors = 314009018,314007924,314005550)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs:
```diff
@@ -2217,5 +2217,35 @@ public void M(string s)
+            var symbolInfo = model.GetSymbolInfo(select);
+
+            Assert.Null(symbolInfo.Symbol);
+            Assert.Empty(symbolInfo.CandidateSymbols);
```
>symbolInfo.CandidateSymbols [](start = 25, length = 27)

I think it would be good to have a test where we can observe that a non-empty set is preserved, including a success case (do we have one already?).

333fred replied:
I'll try to find a repro, but the nature of this bug means, I think, that this would be difficult. The issue was that because there was a single "method" symbol, and semantic info wasn't being correctly preserved, the semantic model thought overload resolution succeeded. If there are multiple candidates, then the semantic model could not have been confused.

---
In reply to: [314007446](https://github.com/dotnet/roslyn/pull/37964#discussion_r314007446) [](ancestors = 314007446)

AlekseyTs commented:
I am curious if there are more "ad-hoc" fields in the bound nodes that are not lazy-calculated.

AlekseyTs commented:
Done with review pass (iteration 1)

333fred commented:
I took a look through. `BoundQueryClause` had some that will be moved in the next PR. Otherwise, all properties looked to be calculated from existing, rather than actually having storage.

---
In reply to: [521358836](https://github.com/dotnet/roslyn/pull/37964#issuecomment-521358836) [](ancestors = 521358836)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [521358949](https://github.com/dotnet/roslyn/pull/37964#issuecomment-521358949) [](ancestors = 521358949)

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -269,7 +269,11 @@
     <Field Name="Operand" Type="BoundExpression"/>
     <Field Name="ConstantValueOpt" Type="ConstantValue?"/>
     <Field Name="MethodOpt" Type="MethodSymbol?"/>
+    <!--The set of method symbols from which this operator's method was chosen.
```
><!--The set of method symbols from which this operator's method was chosen. [](start = 4, length = 75)

Is this supposed to go before OriginalUserDefinedOperatorsOpt?

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -416,7 +432,11 @@
     <Field Name="Right" Type="BoundExpression"/>
     <Field Name="LeftConversion" Type="Conversion"/>
     <Field Name="FinalConversion" Type="Conversion"/>
+    <!--The set of method symbols from which this operator's method was chosen.
```
><!--The set of method symbols from which this operator's method was chosen. [](start = 4, length = 75)

Is this supposed to go before OriginalUserDefinedOperatorsOpt?

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -1614,6 +1642,7 @@
     <Field Name="InvokedAsExtensionMethod" Type="bool"/>
     <Field Name="ResultKind" PropertyOverrides="true" Type="LookupResultKind"/>
     <Field Name="BinderOpt" Type="Binder?" />
+    <Field Name="OriginalMethodsOpt" Type="ImmutableArray&lt;MethodSymbol&gt;" Null="Allow"/>
```
><Field Name="OriginalMethodsOpt" Type="ImmutableArray&lt;MethodSymbol&gt;" Null="Allow"/> [](start = 4, length = 89)

Where did this property come from?

AlekseyTs commented:
Done with review pass (iteration 4)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [521718617](https://github.com/dotnet/roslyn/pull/37964#issuecomment-521718617) [](ancestors = 521718617)

333fred commented:
@dotnet/roslyn-compiler for a second review please.

cston commented on src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs:
```diff
@@ -2217,5 +2217,73 @@ public void M(string s)
         }
+
+        [Fact, WorkItem(37879, "https://github.com/dotnet/roslyn/issues/37879")]
+        public void MissingSymbols_ReinferedParent()
```
>infered [](start = 37, length = 7)

Typo, here and below.

333fred commented:
Even though the status for Windows_Desktop_Unit_Tests debug_32 said failed on GitHub, it's actually passing if you check the `checks` page, and the CI itself is green. The hook just didn't work for status updates.
